torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -625,6 +701,22 @@ void Reducer::finalize_backward() {
       finalize_bucket_dense(bucket);
     }
   }
+
+  // Reset unused parameter accounting.
+  for (size_t i = 0; i < local_used_maps_1_->size(); i++) {
+    (*local_used_maps_1_)[i].fill_(0);
+    (*local_used_maps_2_)[i].fill_(0);
+  }
```

Comment:
This needs to happen after you have waited on the work (in case lazy waiting never happened), because otherwise NCCL may still be mutating this memory until after you have filled it with 0's.